### ISSUE TITLE
Add experimental DuckDB support

### ISSUE THREAD

**Issue by MirkoTeran** [jOOQ/jOOQ#12052]
### Use case:

We are starting to use [DuckDB ](https://duckdb.org/) for some of the smaller clients. Would be nice to have native support for it in jOOQ.

### Possible solution you'd like to see:

Adding DuckDB as supported DB in one of the editions.

### Possible workarounds:

We use SQLite dialect and it works so far as most of our queries are very simple.

### Remarks by @lukaseder

**Bugs and limitations**:

- https://github.com/duckdb/duckdb/issues/6737
- https://github.com/duckdb/duckdb/discussions/7173
- https://github.com/duckdb/duckdb/issues/7118
- https://github.com/duckdb/duckdb/issues/7119
- https://github.com/duckdb/duckdb/issues/7120
- https://github.com/duckdb/duckdb/issues/7168
- https://github.com/duckdb/duckdb/issues/7169
- https://github.com/duckdb/duckdb/issues/7182
- https://github.com/duckdb/duckdb/issues/7183
- https://github.com/duckdb/duckdb/issues/7188
- https://github.com/duckdb/duckdb/issues/7189
- https://github.com/duckdb/duckdb/issues/7206
- https://github.com/duckdb/duckdb/discussions/7207
- https://github.com/duckdb/duckdb/discussions/7208
- https://github.com/duckdb/duckdb/issues/7209
- https://github.com/duckdb/duckdb/issues/7212
- https://github.com/duckdb/duckdb/issues/7214
- https://github.com/duckdb/duckdb/issues/7215
- https://github.com/duckdb/duckdb/issues/7216
- https://github.com/duckdb/duckdb/issues/7218
- https://github.com/duckdb/duckdb/issues/7219
- https://github.com/duckdb/duckdb/issues/7220
- https://github.com/duckdb/duckdb/issues/7230
- https://github.com/duckdb/duckdb/issues/7321
- https://github.com/duckdb/duckdb/issues/8828
- https://github.com/duckdb/duckdb/discussions/8832
- https://github.com/duckdb/duckdb/issues/8836
- https://github.com/duckdb/duckdb/issues/8838

**Features not yet supported**:

- `MACRO`: These act like (scalar or table valued) functions, but without any parameter data types, I suspect they're hard to support in the code generator in a meaningful way. We'll implement things if someone requests it.

### See also:

- https://github.com/jOOQ/jOOQ/discussions/14059

**Comment by lukaseder** [comment]
Thanks for your feature request. We might consider this if there's some traction. Currently, I don't see much of it on https://db-engines.com/en/ranking.

If this is something you'd like to see sooner rather than later, we could discuss a sponsored project. Just let me know.

**Comment by MirkoTeran** [comment]
Great. As long as its noted somewhere that is good. It is not a priority project, even for us.
Thank you.

**Comment by hannesmuehleisen** [comment]
Hey I wrote the DuckDB JDBC driver. Let me know if I can help here.

**Comment by lukaseder** [comment]
> Hey I wrote the DuckDB JDBC driver. Let me know if I can help here.

Thanks for your message. Can you help find budget? :)

**Comment by hannesmuehleisen** [comment]
We certainly can. We can discuss via hannes@cwi.nl

**Comment by lukaseder** [comment]
Thanks, @hannesmuehleisen I'll be in touch

**Comment by shubhamsinha** [comment]
@lukaseder any update on this? We currently also have a need to support jooq with duckdb. Happy to discuss how we can contribute with the budget as well

**Comment by lukaseder** [comment]
Thanks, @shubhamsinha. Can you please reach out to sales@datageekery.com?

**Comment by lukaseder** [comment]
I'm going to look into this for jOOQ 3.19. While this issue has received less love from the community than others, I feel that DuckDB is all the hype on Twitter, so let's be proactive here.

**Comment by asm0dey** [comment]
I believe that it got less love from the community because for most part of users it's just SQLite, they don't really care about analytical functions and capabilities

**Comment by lukaseder** [comment]
I've been documenting a few bugs and limitations:


- https://github.com/duckdb/duckdb/issues/6737
- https://github.com/duckdb/duckdb/discussions/7173
- https://github.com/duckdb/duckdb/issues/7118
- https://github.com/duckdb/duckdb/issues/7119
- https://github.com/duckdb/duckdb/issues/7120
- https://github.com/duckdb/duckdb/issues/7168
- https://github.com/duckdb/duckdb/issues/7169
- https://github.com/duckdb/duckdb/issues/7182
- https://github.com/duckdb/duckdb/issues/7183
- https://github.com/duckdb/duckdb/issues/7188
- https://github.com/duckdb/duckdb/issues/7189
- https://github.com/duckdb/duckdb/issues/7206

As far as jOOQ's integration tests are concerned, I suspect none of these features are really usable yet:

- `INDEX`
- `FOREIGN KEY`
- `UNIQUE` constraints

They're not show stoppers (many RDBMS don't support constraints at all), but I wish I had noticed earlier that constraints and indesxes are *very beta*, still.

**Comment by hannes** [comment]
Thanks so much for taking the time to file those issues!

**Comment by lukaseder** [comment]
Someone has to do it. And it's either me, or the folks at SQLancer 😅

**Comment by lukaseder** [comment]
I've upgraded to the latest snapshots to verify some bug fixes and encountered this crash:

- https://github.com/duckdb/duckdb/issues/7321

While the logical implementation of this new jOOQ dialect is certainly already OK, these types of bugs are show stoppers for an adoption, currently. I'll release the status quo as an experimental dialect in jOOQ 3.19, with a clear disclaimer that jOOQ doesn't recommend to use DuckDB in production yet (the 0.7 release version obviously hints at this as well).

**Comment by lukaseder** [comment]
Had a bit more fun with UDTs and enums in DuckDB today (given that I've been reviewing jOOQ's own UDT support, recently). Found more bugs / missing features:

- https://github.com/duckdb/duckdb/issues/8828
- https://github.com/duckdb/duckdb/discussions/8832
- https://github.com/duckdb/duckdb/issues/8836
- https://github.com/duckdb/duckdb/issues/8838

I guess the dialect will remain in "experimental" state for another while... 😅

**Comment by rkennedy-mode** [comment]
I was trying to kick the tires on 3.19.0-SNAPSHOT and ran into issues trying to parse the following SQL:

```sql
SELECT column_a, column_b, date_diff('month', DATE '1992-09-15', DATE '1992-11-14') FROM table_x
```

I used the following code:

```java
public static void main(String[] args) throws IOException {
    var sqlBytes = Parse.class.getResourceAsStream("/duckdb_simple_query.sql").readAllBytes();
    var sql = new String(sqlBytes, StandardCharsets.UTF_8);

    var query = DSL.using(SQLDialect.DUCKDB)
            .parser()
            .parseQuery(sql);
}
```

The parsing failed with the following exception:

```
Exception in thread "main" org.jooq.impl.ParserException: Unknown function: [1:38] SELECT column_a, column_b, date_diff([*]'month', DATE '1992-09-15', DATE '1992-11-14') FROM table_x

```

I'm curious how far along to expect DuckDB dialect support to be at this point. Am I just too early to be trying this out or is this a genuine bug?

**Comment by lukaseder** [comment]
jOOQ's integration tests aren't open source. I've documented all bugs that I've found in this issue. The segfault doesn't reproduce anymore.

**Comment by elefeint** [comment]
JDBC segfault was fixed in DuckDB 0.9.0 ([issue description](https://github.com/duckdb/duckdb/pull/7909), [fix](https://github.com/duckdb/duckdb/pull/7991))

**Comment by lukaseder** [comment]
Further work will be done via a new issue in the next minor release 3.20:

- https://github.com/jOOQ/jOOQ/issues/15732

I'll publish the experimental support as it is, currently, when 3.19 is published in the next few weeks.
